doc/developer/design/20230802_webhook_idempotency.md

```diff
@@ -0,0 +1,403 @@
+However, that is _not_ desirable for a webhook source, whose domain of keys is
+unreasonably large.
+
+_I have not meaningfully dug into the feasibility of the following approach, so
+suggestions very welcome._
+
+To facilitate this, we could introduce some kind of closure that permits
+mutating upsert state which could take in e.g. all upsert state + an immutable
+reference to the source config.
+
+In the case of webhook source, we could determine the last time we reduced the
+state and every `<period of time>`, we could iterate over the upsert state and
+remove keys we no longer want.
+
+Another option is to set a limit on the amount of state we keep in the upsert
+envelope and evict state based on some closure (e.g. we would want oldest values
+in the case of webhook sources). Tuning this seems annoying, though, because you
+would want to evict batches of records to avoid having to clean up records on
+every insert.
+
+#### Webhook gRPC server
+
+We can set up a new gRPC server that runs in `clusterd` alongside the persist
+and storage server gRPC services.
+
+This server has a few functions:
+- External API
+    - Ingest webhook requests from `environmentd`
+    - Send responses when request durably recorded
+- `clusterd`-internal API
+    - Maintain a set of channels to demultiplex requests into a channel where a
+      running dataflow can ingest it.
+    - Maintain an `mpsc` to` receive responses from dataflows once they've
+      durably recorded the request and send responses to `environemntd`.
+
+The external API can be facilitated through a standard gRPC service
+implementation.
+
+The `clusterd`-internal API requires modifying in-memory state, so needs to
+provide a sender and receiver for requests that can both register and
+un-register `GlobalId`s to bidirectional channels.
+
+In essence, this gRPC server works as a reverse proxy for sending requests to
+Timely workers.
+
+**Early-arriving messages**
+
+Each `clusterd` will spawn a webhook server on boot, just as it does with the
+storage server.
+
+This means that there is possibility of webhook requests coming into the service
+because it's been given a channel to propagate them to. In this case, I suggest
+we just return a status code equivalent to service not available yet, please
+retry later.
+
+**`mz_service`?**
+
+`mz_service`'s module-level documentation says:
+
+> The code shared here is not obviously reusable by future services we may split
+> out of environmentd.
+
+And poking around the code, this seems to be true. Paul and I noodled on how to
+use this module for this service and while it seems possible, it seems more
+legible and simpler to maintain to just roll a new gRPC server.
+
+During the implementation, I plan to use persist's PubSub service as a reference
+implementation and will extract common code from it into some _other_ shard
+module and/or unify the implementation with `mz_service`.
```

Comment:
What goes wrong if we re-use the existing storage commands/responses? It seems natural to me to have a storage command/response like:

```rust
enum StorageCommand {
    // ...
    AppendWebhookData {
        source_id: GlobalId,
        data: // ...
    }
}
```

and a response like:

```rust
enum StorageResponse {
    // ...
    WebhookDataAppended {
        source_id: GlobalId,
        idempotency_keys: Vec<String>,
    }
}
```

If we're worried about mixing control plane and data plane traffic—though I'm not sure how much it matters, because I think the controller needs to perform admission control, not the replica itself—you could add a new RPC to the existing service without creating a new service:

```diff
diff --git a/src/storage-client/src/client.proto b/src/storage-client/src/client.proto
index de3182fd3e..b0ab0e204d 100644
--- a/src/storage-client/src/client.proto
+++ b/src/storage-client/src/client.proto
@@ -23,6 +23,7 @@ package mz_storage_client.client;
 
 service ProtoStorage {
     rpc CommandResponseStream (stream ProtoStorageCommand) returns (stream ProtoStorageResponse);
+    rpc WebhookStream (stream ProtoWebhookCommand) returns (stream ProtoWebhookResponse);
 }
 
 message ProtoCompaction {
```

My thinking here is that this _is_ logically the same service as the existing storage service. It's not like the persist PubSub service that we may want to some day split out of `environmentd`. The webhook service is necessarily part of `clusterd`.

Reply:
One nice thing about using the existing `StorageCommand` and `StorageResponse`s is that you don't have to worry about misorderings. The controller can ensure that it never sends webhook data before creating the webhook source.